### PR TITLE
Implementing prognostics horizon metric for prog_algs.

### PR DESCRIPTION
**Pull request by lawrence-hwang**
PR recreated due to branch renaming. Implementing prognostics horizon metric for prog_algs.

### CODE REVIEW

File: tests/test_metrics.py [hunk 1]
```diff
@@ -391,6 +391,35 @@ def test_toe_profile_metrics(self):
...
+                UnweightedSamples(data)  # ToE Prediction
+            )
+
+        # Define test sample ground truth
+        GROUND_TRUTH = {'a': 9.0, 'b': 8.0, 'c': 18.0}
+        # Create rudimentary criteria equation
+        def criteria_eqn(toe : ToEPredictionProfile, ground_truth : dict):
+            result = {}
+            for key, value in ground_truth.items():
+                if abs(toe.mean[key] - value < 0.1):
```
Comment (teubert) [inline tests/test_metrics.py:413]:
```suggestion
                if abs(toe.mean[key] - value) < 0.1:
```

File: tests/test_metrics.py [hunk 2]
```diff
@@ -391,6 +391,35 @@ def test_toe_profile_metrics(self):
...
+                    result[key] = True
+                else:
+                    result[key] = False
+            return result
+        from prog_algs.metrics import prognostic_horizon
+        # Prognostic horizon metric testing
+        ph_metrics = prognostic_horizon(profile, criteria_eqn, GROUND_TRUTH)
+        self.assertDictEqual(ph_metrics, {'c':9.0})
+
```
Comment (teubert) [inline tests/test_metrics.py:422]: Other tests we want:
* Where all criteria are met at different times
* Where 2 criteria are met at once
* Where at least one criteria is met at beginning of the prediction

**Review by teubert — CHANGES_REQUESTED** [review]

File: src/prog_algs/metrics/toe_profile_metrics.py [hunk 3]
```diff
@@ -41,3 +42,33 @@ def alpha_lambda(toe_profile : ToEPredictionProfile, ground_truth : dict, lambda
...
+    for (t_prediction, toe) in toe_profile.items():
+        criteria_eqn_dict = criteria_eqn(toe, ground_truth) # -> dict[event_names as str, bool]
+        for k,v in criteria_eqn_dict.items():
+            if v: # eqn dict value is True
+                if not ph_first_met[k]:
+                    ph_result[k] = ground_truth[k] - t_prediction # PH = EOL - ti # ground truth is a dictionary {'EOD': 3005.2} should be ph_result[k] = g_truth[key] - t_prediction
+                    ph_first_met[k] = True
+                    if (ph_result.keys() == ph_first_met.keys()) and (all(v for v in ph_first_met.values())):
+                        return ph_result
+
```
Comment (teubert) [inline src/prog_algs/metrics/toe_profile_metrics.py:74]: Return none for each event not met

File: src/prog_algs/metrics/toe_profile_metrics.py [hunk 4]
```diff
@@ -41,3 +43,42 @@ def alpha_lambda(toe_profile : ToEPredictionProfile, ground_truth : dict, lambda
                     print('\ttoe:', toe.key(key))
                     print('\tBounds: [{} - {}]({}%)'.format(lower_bound, upper_bound, toe.percentage_in_bounds([lower_bound, upper_bound])[key]))
             return result
+
+def prognostic_horizon(toe_profile : ToEPredictionProfile, criteria_eqn : Callable, ground_truth : dict, **kwargs):
+    """
+    Compute prognostic horizon metric, given by the difference between a time ti, when the predictions meet specified performance criteria, and the time corresponding to the end of life (EoL).
```
Comment (teubert) [inline src/prog_algs/metrics/toe_profile_metrics.py:49]:
```suggestion
    Compute prognostic horizon metric, defined as the difference between a time ti, when the predictions meet specified performance criteria, and the time corresponding to the true Time of Event (ToE), for each event.
```

File: src/prog_algs/metrics/toe_profile_metrics.py [hunk 5]
```diff
@@ -41,3 +43,42 @@ def alpha_lambda(toe_profile : ToEPredictionProfile, ground_truth : dict, lambda
...
+    PH = EOL - ti
```
Comment (teubert) [inline src/prog_algs/metrics/toe_profile_metrics.py:50]:
```suggestion
    PH = ToE - ti
```

File: src/prog_algs/metrics/toe_profile_metrics.py [hunk 6]
```diff
@@ -41,3 +43,42 @@ def alpha_lambda(toe_profile : ToEPredictionProfile, ground_truth : dict, lambda
...
+    Args:
+        toe_profile (ToEPredictionProfile): A profile of predictions, the combination of multiple predictions
+        criteria_eqn (Callable function): A function calculating whether ground truth meets some criteria for corresponding key in ToEPredictionProfile
```
Comment (teubert) [inline src/prog_algs/metrics/toe_profile_metrics.py:53]:
```suggestion
        criteria_eqn (Callable function): A function (toe: UncertainData, ground_truth: dict[str, float]) -> dict[str, bool] calculating whether a prediction in ToEPredictionProfile meets some criteria. \n
            | Args: 
            |  * toe (UncertainData): A single prediction of Time of Event (ToE)
            |  * ground truth (dict[str, float]): Ground truth passed into prognostics_horizon
            | Returns: Map of event names to boolean representing if the event has been met. 
            |   e.g., {'event1': True, 'event2': False}
```

File: src/prog_algs/metrics/toe_profile_metrics.py [hunk 7]
```diff
@@ -41,3 +43,42 @@ def alpha_lambda(toe_profile : ToEPredictionProfile, ground_truth : dict, lambda
...
+        ground_truth (dict): Dictionary containing ground truth; specified as key, value pairs for event and its value
```
Comment (teubert) [inline src/prog_algs/metrics/toe_profile_metrics.py:54]:
```suggestion
        ground_truth (dict): Dictionary containing ground truth; specified as key, value pairs for event and its value. E.g, {'event1': 47.3, 'event2': 52.1, 'event3': 46.1}
```

File: src/prog_algs/metrics/toe_profile_metrics.py [hunk 8]
```diff
@@ -41,3 +43,42 @@ def alpha_lambda(toe_profile : ToEPredictionProfile, ground_truth : dict, lambda
...
+        kwargs (optional): configuration arguments. Accepted args include:
+            * keys (list[string], optional): list of keys to use. If not provided, all keys are used.
```
Comment (teubert) [inline src/prog_algs/metrics/toe_profile_metrics.py:56]: We don't use keys configuration argument. Instead, we just use print. Please replace this with a description of print

File: src/prog_algs/metrics/toe_profile_metrics.py [hunk 9]
```diff
@@ -41,3 +43,42 @@ def alpha_lambda(toe_profile : ToEPredictionProfile, ground_truth : dict, lambda
...
+
+    Returns:
+        dict: Dictionary containing prognostic horizon calculations (value) for each event (key)
```
Comment (teubert) [inline src/prog_algs/metrics/toe_profile_metrics.py:59]:
```suggestion
        dict: Dictionary containing prognostic horizon calculations (value) for each event (key). e.g., {'event1': 12.3, 'event2': 15.1}
```

File: src/prog_algs/metrics/toe_profile_metrics.py [hunk 10]
```diff
@@ -41,3 +43,42 @@ def alpha_lambda(toe_profile : ToEPredictionProfile, ground_truth : dict, lambda
...
+    """
+    params = {
+        'print': False
+    }
+    params.update(kwargs)
+
+    ph_result = {k:False for k in ground_truth.keys()} # False means not yet met; will be either a numerical value or None if met
```
Comment (teubert) [inline src/prog_algs/metrics/toe_profile_metrics.py:66]:
```suggestion
    ph_result = {k:None for k in ground_truth.keys()} # False means not yet met; will be either a numerical value or None if met
```
Why not start with None and change the comparisons below to "is None" or "is not None"

Comment (teubert) [reply src/prog_algs/metrics/toe_profile_metrics.py:66]: Then we should be able to remove lines 80-82

File: src/prog_algs/predictors/toe_prediction_profile.py [hunk 11]
```diff
@@ -61,3 +61,19 @@ def alpha_lambda(self, ground_truth : Dict[str, float], lambda_value : float, al
         """
         from ..metrics import alpha_lambda
         return alpha_lambda(self, ground_truth, lambda_value, alpha, beta, **kwargs)
+
+    def prognostic_horizon(self, toe_profile, criteria_eqn, ground_truth, **kwargs) -> Dict[str, float]:
+        """
+        Compute prognostic horizon metric, given by the difference between a time ti, when the predictions meet specified performance criteria, and the time corresponding to the end of life (EoL).
```
Comment (teubert) [inline src/prog_algs/predictors/toe_prediction_profile.py:67]: Propagate changes from above.

File: tests/test_metrics.py [hunk 12]
```diff
@@ -391,6 +391,50 @@ def test_toe_profile_metrics(self):
...
+        # Test 4: at least one criteria is met at beginning of the prediction
+        GROUND_TRUTH = {'a': 10, 'b': 8.0, 'c': 10.0}
+        self.assertDictEqual(prognostic_horizon(profile, criteria_eqn, GROUND_TRUTH), {'a': 1.0, 'b': None, 'c': None})
+        # Test 5: criteria not met for any
+        GROUND_TRUTH = {'a': 9.0, 'b': 8.0, 'c': 8.0}
+        self.assertDictEqual(prognostic_horizon(profile, criteria_eqn, GROUND_TRUTH), {'a': None, 'b': None, 'c': None})
+
 # This allows the module to be executed directly    
 def run_tests():
     l = unittest.TestLoader()
```
Comment (teubert) [inline tests/test_metrics.py:440]: Make sure you add to the playback example as well

**Review by teubert — CHANGES_REQUESTED** [review]
A few changes- mostly documentation. It's almost there!

**Comment by lawrence-hwang** [comment]
Screen Shot 2022-04-15 at 1 25 23 PM

**Comment by lawrence-hwang** [comment]
example: Given a t_p of 100, a ground truth of 500, and predictions of type unweighted samples of 445, 470, 495, 520, and 545, if the beta is 0.6 (i.e., 60% must be within alpha bounds), an alpha of 0.1 would be met. Because true TtE is 400, and 60% is within 40 (i.e. 0.1 * 400), but an alpha of 0.05 would not be met, because only 40% is within 20 of 500

**Review by teubert — APPROVED** [review]

**Pull request merged**